doc/general_design.md

```diff
@@ -4,31 +4,79 @@
 
 > Implement a prototype job worker service that provides an API to run arbitrary Linux processes.
 
-## General Design ##
+## Design ##
 
 This project will have client app and server daemon app. Server app will open API and receive the commands from client side then run the command as the client wish.
 
 Client side is a command line app for users run the commands through terminal. Users should have the ability of `start, stop, get status, and stream output of a job`.
 
-### runtime ###
+### Server side ###
 
+Server side application should listen the port `9090` for gRPC requests. 
 
+#### runtime ####
 
-### API ###
+This project will use [tokio 1.x](https://docs.rs/tokio/latest/tokio/) as the runtime for supporting async features.
 
-gRPC
+Tokio runtime will handle each the function of each requests. As the "backend" after gRPC api.
 
-### CLI interface ###
+#### API ####
 
-### Authorization ###
+This project server side will open gRPC API for other service/cli/etc.. It uses [tonic](https://docs.rs/tonic/latest/tonic/) as the gRPC lib for code generating and server running.
 
-mTCP
+gRPC API schema includes the fields of gRPC request and response for client and server side. 
 
-### cgroups ###
+```grpc
+proto Request {
+    command ;; start/stop/get status/stream output/etc. commands
+    client authorization info {
+        ...
+    }
+}
 
+proto Response {
+    command_status ;; success/failed/etc. for the API call
+    job_statuc {
+        job_status ;; status of job
+        ...
+    }
+}
+
+```
+
+#### Authorization ####
+
+Use mTLS as the authentication method for making sure the identity of client. After the transport layer (mTLS) authorization, we also need a way to know which client it is. In this project, I might hard code some `client id` of clients, keep the publickey-client relations inside server side (which should keep in DB in real world). 
```

Comment:
What will you use to determine the identity of the client?

Reply:
Followup question: what is the authorization scheme, once you have a client id/name? Are you going to restrict the ability to create new jobs to certain users or groups? Are jobs tied to users?

Reply:
this is the part I am not sure about. My thought is to put the client_id in the RPC request, and store client_ids-group-permission relations on the server side. It gonna make this project easier. But it has risks in the real world:

People who have legal crt and pem can fake their client id to check other jobs that they do not have the permission of. 

then I think maybe the server can generate a key pair for every client, the client uses the public key to encrypt the data and send the encrypted data to the server, then server

Reply:
Isn't there something that you're already using and that the client can't fake, that could be used for this purpose?

Reply:
Do you mean the public key, which is the pair of the private key, can be used to check if the client also has the private key or not?